### ISSUE TITLE
DropwizardSSLConnectionSocketFactory not honoring custom provider?

### ISSUE THREAD

**Issue by nheitz** [dropwizard/dropwizard#2366]
```
   private static KeyStore loadKeyStore(String type, File path, String password) throws Exception {
        final KeyStore keyStore = KeyStore.getInstance(type);
        try (InputStream inputStream = new FileInputStream(path)) {
            keyStore.load(inputStream, password.toCharArray());
        }
        return keyStore;
    }
```
Using : Dropwizard 1.3.2

The above code is used to load the trust store associated with the tls config for an SSL client.  However, this ignores the provider (eg, "BC") that might be required to read a PKCS12 store properly.

I believe this should be 
```  
final KeyStore keyStore = KeyStore.getInstance(type, "<provider as String>");
```

Note:  using "BC" as a provider works on the server side when listening on an SSL port.

EDIT:

I've discovered a "workaround", which involves hardwiring BouncyCastle as a higher priority provider to the Sun JCE.  Whilst this works, it immediately imposes the performance penalty of using BouncyCastle where optimised AES is better in the SunJCE.  Whilst it could be possible to go to every location and specify the provider by name, I believe the better solution is to offer the truststore/keytore provider option for the dropwizard tls client.

**Comment by arteam** [comment]
Interesting observation! I think you are right, and we should load the keystore based on the provided security provider in the configuration. The biggest questions for me is that `KeyStore getInstance(String type)` traverses all the registered providers looking for one that supports the specified type, whilst `KeyStore getInstance(String type, String provider)` checks only the specified provider. But I assume users who specify BouncyCastle and Conscrypt support different keystore types.

**Comment by nheitz** [comment]
@arteam ...the only thing that worries me slightly is that BouncyCastle does not have a provider for JSSE, only JCE (it seems).  When I tried to set "provider" to "BC" in the tls config, I ran into that issue.  What I am unsure about is whether the code I highlighted above is used also for provisioning the SSE implementation.

**Comment by arteam** [comment]
Right, so it shouldn't be possible to use TLS with BC. From the other hand, Conscrypt is a JSSE provider, but it not terribly useful for loading keystores.

**Comment by arteam** [comment]
It would be nice to get a reproducer of the issue with a custom provider.

**Comment by nheitz** [comment]
I can provide you with a pkcs12 truststore that only works with BC, or I can try out your did myself... Which would be preferable?

**Comment by arteam** [comment]
Either way is fine. If you would like to take a swing on the bugfix, go ahead!

**Comment by nheitz** [comment]
I was seriously tempted to have a go....but then I faced a moment of uncertainty when I saw how you'd approached it.  I will attach an exemplar pkcs12 keystore that should fail under SunJCE (java 8) and succeed with BC.

In looking at your implementation, it looks like you are reusing "configuration.getProvider" for specifying "BC" for trust store.  When I tried to set "provider" to "BC" in configuration, however, dropwizard failed to create the secure socket extension, because BC doesn't have an implementation of that.  

It is a bit tricky to see where that use of the provider is...would that failure somehow be avoided by your try-catch fallback approach?

[keystore-trustore.zip](https://github.com/dropwizard/dropwizard/files/2066308/keystore-trustore.zip)

**Comment by nheitz** [comment]
It is line 80-83 of DropwizardSSLConnectionSocketFactory that I believe will continue to cause failure.  

My original idea for a fix was to add another property to the tlsConfig called "trustStoreProvider", possibly...but then it's not clear to me what the "provider" property is intended to be used for.  Would it be overengineering to add "sseProvider", "keystoreProvider", and "truststoreProvider" to the tlsConfiguration, for completeness?

**Comment by nheitz** [comment]
I think there might be additional evidence to support splitting the providers in dropwizard config:
it turns out there **is** a BouncyCastle TLS provider, but it has a different provider name, so having a single "provider" attribute in yaml (for tlsConfig) used for both SSE and JCE is problematic.
This flirts with being a backwards incompatible change, depending on naming, so I await your thoughts, @arteam .

**Comment by nheitz** [comment]
@arteam 
Ok, I had a go at the solution.
https://github.com/dropwizard/dropwizard/pull/2390

Not too onerous in the end.  I did add bouncycastle as test component explicitly to dropwizard-client, as it is impossible to do effective testing otherwise.  I did not modify the BOM....please let me know if I should update the BOM as well to inherit version information for test dependencies, or if what I've done is sufficient.

**Comment by arteam** [comment]
Thanks, I will take a look.

**Comment by nheitz** [comment]
@arteam ,  Sorry for the newbie question, but as this is my first PR to dropwizard....do I close this issue, given that the PR has been merged, or do you do that at some point?

**Comment by arteam** [comment]
Either way works for me! I can do it manually, but usually issues get closed automatically by commit messages from pull requests.

**Comment by arteam** [comment]
Closed via #2390
